---
title: "day2_exercise"
output: html_document
date: "2024-07-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###### Course 22123: Computational precision medicine
###### Day 2 practical: working with gene expression data in R
###### By Lars Ronn Olsen
###### 02/06/2023
###### Technical University of Denmark

#### Load necessary packages (you can back fill once you know what packages you will use)
```{r}
library(tidyverse)
```


#### Load expression data from Xena ("gene expression RNAseq" - "HTSeq - Counts")
```{r}
gbm_counts <- read.table("../00_raw_data/TCGA-GBM.htseq_counts.tsv", header = TRUE, row.names = 1)
```


#### Un-"log2" the data
```{r}
gbm_counts <- round((2^gbm_counts)-1, digits = 0)
```


#### Check out the distribution of the expression of a sample or two. What kind of distribution are you looking at?
```{r}
ggplot(data = gbm_counts, 
       mapping = aes(x = TCGA.26.5135.01A)) + 
  geom_density(fill="#69b3a2", color="#e9ecef", alpha=0.8) +
  ggtitle("Distribution of gene expression for sample TCGA.06.0878.01A")
```


#### Load expression data from Xena ("gene expression RNAseq" - "HTSeq - FPKM") - you should be working with this from now on
```{r}
gbm_fpkm <- read.table("../00_raw_data/TCGA-GBM.htseq_fpkm.tsv", header = TRUE, row.names = 1)
```


#### Check out the distribution of the expression of a sample or two. What changed? Why should we transform the raw counts?
```{r}
ggplot(data = gbm_fpkm, 
       mapping = aes(x = TCGA.26.5135.01A)) + 
  geom_density(fill="#69b3a2", color="#e9ecef", alpha=0.8) +
  ggtitle("Distribution of gene expression for sample TCGA.06.0878.01A")
```


#### Load phenotype data from Xena ("phenotype" - "phenotype")
```{r}
gbm_pheno <- read.table("../00_raw_data/TCGA-GBM.GDC_phenotype.tsv", header = TRUE, sep = "\t", quote="")
```


#### Take a closer look at the column called "sample_type.samples" and read here: https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes
```{r}
table(gbm_pheno$sample_type.samples)
```


#### Take a closer look at the column called "batch_number"
```{r}
table(gbm_pheno$batch_number)
```


#### Extract both those variables from the phenotype table along with the column called "submitter_id.samples".
```{r}
gbm_pheno <- gbm_pheno %>% 
              select(c("submitter_id.samples", "batch_number", "sample_type.samples"))
```


#### Do a little cleanup:
#### Rename sample type ids to match those in your expression table columns
```{r}
gbm_pheno <- gbm_pheno %>% 
  mutate(across("submitter_id.samples", str_replace, '-', '.'))
```


#### Subset the phenotype table to contain only the samples you have expression data for
```{r}
gbm_pheno <- gbm_pheno %>% 
  filter(submitter_id.samples %in% colnames(gbm_fpkm))
```


#### Optional: save Rdata object for later use
```{r}
save(gbm_counts, gbm_fpkm, gbm_pheno, file = "../01_tidy_data/gbm.Rdata")
```


#### Calculate PCA on count matrix
```{r, eval=FALSE}
pca_count <- prcomp(t(gbm_counts), scale. = TRUE)
```

#### Why did this not work?
It is because some of the genes have zero variance. 

#### Solve the problem and try again
```{r}
gbm_counts <- gbm_counts %>% 
                filter(rowSums(gbm_counts) > 0)

pca_count <- prcomp(t(gbm_counts), scale. = TRUE)
```


#### Check out how much variance is explained
```{r}
summary(pca_count)
```

#### Is this more or less than you would expect?
Less than expected. PC1 and PC2 only 0.108 and 0.05813 respectively. 

#### Plot the first two principal components of the PCA
```{r}
library(broom)

pca_count %>% 
  augment(t(gbm_counts)) %>% 
  ggplot(aes(x = .fittedPC1, y = .fittedPC2)) +
  geom_point() +
  ggtitle("PC1 vs PC2 on Counts")
```


#### Now try the same with FPKM transformed data
```{r}
gbm_fpkm <- gbm_fpkm %>% 
              filter(rowSums(gbm_fpkm) > 0)

pca_fpkm <- prcomp(t(gbm_fpkm), scale. = TRUE)

summary(pca_fpkm)

pca_fpkm %>% 
  augment(t(gbm_fpkm)) %>% 
  ggplot(aes(x = .fittedPC1, y = .fittedPC2)) +
  geom_point() +
  ggtitle("PC1 vs PC2 on FPKM")

```

#### Do you get the same results in terms of variance explained and visualization of PC1 and PC2? Why/why not? Which do you think is better to use?
No. 

#### Plot PCA of FPKM and color by batch
```{r}
pca_fpkm %>% 
  augment(t(gbm_fpkm)) %>% 
  left_join(gbm_pheno, by = c(".rownames" = "submitter_id.samples")) %>% 
  ggplot(aes(x = .fittedPC1, y = .fittedPC2, colour = batch_number)) +
  geom_point() +
  ggtitle("PC1 vs PC2 on FPKM colored by batch")
```


#### Do you anticipate batch effects in this dataset?
No

#### Plot PCA and color by sample type
```{r}
pca_fpkm %>% 
  augment(t(gbm_fpkm)) %>% 
  left_join(gbm_pheno, by = c(".rownames" = "submitter_id.samples")) %>% 
  ggplot(aes(x = .fittedPC1, y = .fittedPC2, colour = sample_type.samples)) +
  geom_point() +
  ggtitle("PC1 vs PC2 on FPKM colored by sample type")
```
#### How does this look?
Solid tissue normal is clearly distiguished from primary tumor and recurrent tumor. 

#### Make a heatmap of the genes in the list "diff_genes.txt" (https://teaching.healthtech.dtu.dk/22123/diff_genes.txt)
```{r}
diff_genes <-read.table(url("https://teaching.healthtech.dtu.dk/22123/diff_genes.txt"))

selected_fpkm <- gbm_fpkm %>% 
                  filter(rownames(gbm_fpkm) %in% diff_genes$V1)

heatmap(as.matrix(selected_fpkm), scale = "row", Colv = NA, Rowv = NA)

```


#### Do the clusters in the dendrogram correspond to subtypes or batches?
```{r}
heatmap(as.matrix(selected_fpkm), scale = "row")
```


#### Calculate the correlation between genes and visualize a low, medium, and high correlation pair in three separate scatter plots
#### Calculate the correlation between the genes "ENSG00000249568.1" and "ENSG00000249186.1"
```{r}
t(gbm_fpkm) %>% 
  ggplot(mapping = aes(x = ENSG00000249568.1, y = ENSG00000249186.1)) + 
  geom_point() + 
  geom_smooth(method = lm, se=FALSE) + 
  ggtitle("Correlation plot between genes ENSG00000249568.1 and ENSG00000249186.1")
```


#### Visualize the expression of the gene "ENSG00000267112.1 in primary vs normal samples using boxplots
```{r}
gbm_fpkm %>% 
  filter(rownames(gbm_fpkm) == "ENSG00000267112.1") %>% 
  pivot_longer(cols = contains("TCGA"), 
               names_to = "sample_id", 
               values_to = "gene_expression") %>% 
  left_join(gbm_pheno, by = c("sample_id" = "submitter_id.samples")) %>% 
  ggplot(mapping = aes(x = as.factor(sample_type.samples), y = gene_expression)) + 
  geom_boxplot(fill="slateblue", alpha=0.2) + 
  xlab("Sample Type") +
  ylab("Gene expression (FPKM)") + 
  ggtitle("ENSG00000267112.1 Gene expression in different sample type")
  
```




















